OldSneerJaw/couchster
# Issue 25: Regular expression pattern constraint for attachment references

# Description

When included in a document definition, the `attachmentReference` validation type's `regexPattern` constraint verifies that the value conforms to the specified regular expression pattern. It overrides the value of the document-wide `attachmentConstraints.filenameRegexPattern` constraint for that attachment reference.

# Testing

First, verify that the constraint works on its own (i.e. without being forced to interact with `attachmentConstraints.filenameRegexPattern`):

1. Generate a validation function from `attachment-reference-doc-definitions.js`:

```bash
./make-validation-function -j test/resources/attachment-reference-doc-definitions.js build/test-validation-function.txt
```

2. Copy and paste the validation function into a CouchDB design document
3. Attempt to create a test document with an attachment reference that violates the regex pattern (it should be rejected with 403 Forbidden):

```
PUT /test/attachment-ref-regex-test-1 HTTP/1.1
Host: localhost:5984
Content-Type: application/json
Authorization: Basic ********

{
    "type": "attachmentReferencesDoc",
    "staticRegexPatternValidationProp": "12345.pdf"
}
```

4. Create a test document with an attachment reference that satisfies the regex pattern (it should succeed with 201 Created):

```
PUT /test/attachment-ref-regex-test-1 HTTP/1.1
Host: localhost:5984
Content-Type: application/json
Authorization: Basic ********

{
    "type": "attachmentReferencesDoc",
    "staticRegexPatternValidationProp": "abc.txt"
}
```

Now, verify that the constraint correctly takes precedence over the document-wide `attachmentConstraints.filenameRegexPattern` constraint:

1. Generate a validation function from `attachment-constraints-doc-definitions.js`:

```bash
./make-validation-function -j test/resources/attachment-constraints-doc-definitions.js build/test-validation-function.txt
```

2. Copy and paste the validation function into a CouchDB design document.
3. Create a document:

```
PUT /test/attachment-ref-regex-test-2 HTTP/1.1
Host: localhost:5984
Content-Type: application/json
Authorization: Basic ********

{
    "type": "staticRegularAttachmentsDoc",
    "attachmentRefProp": "abcdef.foo"
}
```

4. Attempt to add an attachment that violates the `attachmentConstraints.filenameRegexPattern` constraint and isn't referenced by the `attachmentReference` property (should be rejected with 403 Forbidden):

```
PUT /test/attachment-ref-regex-test-2/123.txt?rev=1-63fbb434897336bf340d1d8b0460a6d8 HTTP/1.1
Host: localhost:5984
Content-Type: text/plain
Authorization: Basic ********

The file name doesn't match attachmentConstraints.filenameRegexPattern: /^(foo|ba[rz]|qux)\.[a-z]+$/
```

5. Add an attachment that violates the `attachmentConstraints.filenameRegexPattern` constraint but is allowed anyway because it is referenced by the `attachmentReference` property which defines a more permissive `regexPattern` constraint (should be allowed with 201 Created):

```
PUT /test/attachment-ref-regex-test-2/abcdef.foo?rev=1-63fbb434897336bf340d1d8b0460a6d8 HTTP/1.1
Host: localhost:5984
Content-Type: text/plain
Authorization: Basic ********

Matches!
```
# Related Issue

* GitHub issue link: #25
